cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/runtime/monitor/AbstractServiceRoutingHandler.java

```diff
@@ -184,7 +184,10 @@ private HttpURLConnection openConnection(HttpRequest request, String service,
       throw new ServiceUnavailableException(service);
     }
 
-    URI uri = URIScheme.createURI(discoverable, path);
+    // The createURI method accepts a format string while the path string we have
+    // doesn't require any substitution. So we escape any % characters which may be present
+    // due to URL encoding.
+    URI uri = URIScheme.createURI(discoverable, path.replace("%", "%%"));
```

Comment:
I don't think we should do a replace like this, feeling unsafe to me. If the path is url encoded, it's better to do a proper `URLDecoder.decode` of the incoming path. BTW, why the incoming path is encoded? If we get it from the http request `uri()` method, is it not decoded already?

Reply:
I don't think it's because of URI encoding, it's because `createURI` uses String.format. Instead I would do `createURI(discoverable, "%s", path)` as this would show intent better

Reply:
It would make sense also for createURI to at least issue a warning if it has no parameters

Reply:
@chtyim The query parameters in the url aren't decoded. The added unit test fails without the changes.

@tivv  `createURI(discoverable, "%s", path)` is much better, should have thought of that! Made the change. Also added a warn log when the substitution arguments are missing.